app/questionnaire/questionnaire_schema.py

```diff
@@ -839,17 +882,70 @@ def _populate_when_rules_section_dependencies(self) -> None:
             when_rules = self.get_values_for_key(section, "when")
             rules: list = list(when_rules)
 
-            if rules_section_dependencies := self._get_rules_section_dependencies(
-                section["id"], rules
-            ):
+            (
+                rules_section_dependencies,
+                rule_section_dependencies_for_progress_value_source,
+                rule_block_dependencies_for_progress_value_source,
+            ) = self._get_rules_section_dependencies(section["id"], rules)
+
+            if rules_section_dependencies:
                 self._when_rules_section_dependencies_by_section[
                     section["id"]
                 ] = rules_section_dependencies
 
+            if rule_section_dependencies_for_progress_value_source:
+                self._when_rules_section_dependencies_by_section_for_progress_value_source[
+                    section["id"]
+                ] = rule_section_dependencies_for_progress_value_source
+
+            if rule_block_dependencies_for_progress_value_source:
+                self._when_rules_block_dependencies_by_section_for_progress_value_source[
+                    section["id"]
+                ] = rule_block_dependencies_for_progress_value_source
```

Comment:
To avoid repetition/reduce the lines, you can:
```suggestion
            for rule_dependency, when_rules in zip(
                self._get_rules_section_dependencies(section["id"], rules),
                [
                    self._when_rules_section_dependencies_by_section,
                    self._when_rules_section_dependencies_by_section_for_progress_value_source,
                    self._when_rules_block_dependencies_by_section_for_progress_value_source,
                ],
            ):
                if rule_dependency:
                    when_rules[section["id"]] = rule_dependency
```
(checked in unit tests)